### ISSUE TITLE
What is the meaning of return value "z_e, emb" of CVAE_imagenet? What corresponding Loss Function should be used?

### ISSUE THREAD

**Issue by chuchuhe** [kai-wen-yang/CD-VAE#4]
![image](https://user-images.githubusercontent.com/77535209/157657638-8ceecbcd-5918-4710-bde1-77ab75d8d7c9.png)

**Comment by kai-wen-yang** [comment]
> ![image](https://user-images.githubusercontent.com/77535209/157657638-8ceecbcd-5918-4710-bde1-77ab75d8d7c9.png)

As said in the paper, we use VQ-VAE for imageNet which is of high-resolution. So the code corresponds to the loss in VQ-VAE

**Comment by kai-wen-yang** [comment]
> 

The detail of VQ-VAE's loss can be found in Equation (3) in https://arxiv.org/pdf/1711.00937.pdf.
It likes
```
 L_VQVAE = F.mse_loss(gx, x) + torch.mean(torch.norm((emb - z_e.detach())**2, 2, 1)) + 0.5*torch.mean(torch.norm((emb.detach() - z_e)**2, 2, 1))
```

**Comment by GANG370** [comment]
add l1 = F.mse_loss(gx, x) is ok, Could you provide a better loss weight coefficient? I found that the experimental effect was not good according to 1:1

**Comment by kai-wen-yang** [comment]
> add l1 = F.mse_loss(gx, x) is ok, Could you provide a better loss weight coefficient? I found that the experimental effect was not good according to 1:1
Try:
```
L_VQVQE=F.mse_loss(gx, x) + 0.1*[ torch.mean(torch.norm((emb - z_e.detach())**2, 2, 1)) + 0.5*torch.mean(torch.norm((emb.detach() - z_e)**2, 2, 1))]
```

**Comment by GANG370** [comment]
Thank you for your reply, but i need a more loss,that is cross_entropy = lam * F.cross_entropy(out, y[0]) + (1. - lam) * F.cross_entropy(out, y[1]), Please give me another set of coefficients. Thank you very much!

**Comment by kai-wen-yang** [comment]
> Thank you for your reply, but i need a more loss,that is cross_entropy = lam * F.cross_entropy(out, y[0]) + (1. - lam) * F.cross_entropy(out, y[1]), Please give me another set of coefficients. Thank you very much!

A small weight for cross_entropy term is enough, so try:

```
L_VQVAE+0.1*cross_entropy
```

**Comment by GANG370** [comment]
I did the experiment according to the weight you said, and found that with the increase of eopoches, the reconstruction effect is getting worse and worse, almost black, and the rec acc of the test is a very large negative number (such as: -2776), However, the reconstruction loss during training is very small, and the final classification accuracy is not high (75% for the second classification).：I want to ask what is the cause of this, whether it is related to the weight, then how should I adjust it?

**Comment by kai-wen-yang** [comment]
> I did the experiment according to the weight you said, and found that with the increase of eopoches, the reconstruction effect is getting worse and worse, almost black, and the rec acc of the test is a very large negative number (such as: -2776), However, the reconstruction loss during training is very small, and the final classification accuracy is not high (75% for the second classification).：I want to ask what is the cause of this, whether it is related to the weight, then how should I adjust it?

Why the reconstruction loss is small but the rec acc of the test is bad? I think there may exist bug in your VQVAE. I sugget you remove the classifier in your code and train a pure VQVAE using your code, and see what will happen.

**Comment by GANG370** [comment]
> > add l1 = F.mse_loss(gx, x) is ok, Could you provide a better loss weight coefficient? I found that the experimental effect was not good according to 1:1
> > Try:
> 
> ```
> L_VQVQE=F.mse_loss(gx, x) + 0.1*[ torch.mean(torch.norm((emb - z_e.detach())**2, 2, 1)) + 0.5*torch.mean(torch.norm((emb.detach() - z_e)**2, 2, 1))]
> ```
**Why are these two things equal is there something wrong with the loss function here?**  @kai-wen-yang

**Comment by GANG370** [comment]
> > add l1 = F.mse_loss(gx, x) is ok, Could you provide a better loss weight coefficient? I found that the experimental effect was not good according to 1:1
> > Try:
> 
> ```
> L_VQVQE=F.mse_loss(gx, x) + 0.1*[ torch.mean(torch.norm((emb - z_e.detach())**2, 2, 1)) + 0.5*torch.mean(torch.norm((emb.detach() - z_e)**2, 2, 1))]
> ```
**Why do the last two terms of Loss always have the same value?  Is there something wrong with the loss function here?**  @kai-wen-yang

**Comment by kai-wen-yang** [comment]
Have you read https://arxiv.org/pdf/1711.00937.pdf?

**Comment by kai-wen-yang** [comment]
> > > add l1 = F.mse_loss(gx, x) is ok, Could you provide a better loss weight coefficient? I found that the experimental effect was not good according to 1:1
> > > Try:
> > 
> > 
> > ```
> > L_VQVQE=F.mse_loss(gx, x) + 0.1*[ torch.mean(torch.norm((emb - z_e.detach())**2, 2, 1)) + 0.5*torch.mean(torch.norm((emb.detach() - z_e)**2, 2, 1))]
> > ```
> 
> **Why do the last two terms of Loss always have the same value? Is there something wrong with the loss function here?** @kai-wen-yang

Have you read  https://arxiv.org/pdf/1711.00937.pdf?
